### DISCUSSION TITLE
Unable to render hooks nested inside `InstantSearchSSRProvider` 🙏

### DISCUSSION THREAD

**Discussion by Jezfx** [algolia/instantsearch#5665] [Q&A] [answered]
Any help with this would be HUGELY appreciated. I've been super blocked by this 🙏 thank you! 

I'm trying to get SSR to work with next. I've been able to follow the [official next example ](https://github.com/algolia/instantsearch/tree/master/examples/react-hooks/next) and everything works ok until I add a hook into the mix. 

For our app we—effectively—need make a call to `react-query` inside the product card. However, what seems to happen is when InstantSearchSSRProvider is rendering it on the server it's unable to wrap that hook with `QueryClientProvider` in `_app.js`. 

This is the error I'm getting: 
 
<img width="1083" alt="Screenshot 2023-06-05 at 15 12 49" src="https://github.com/algolia/instantsearch/assets/356174/3ebdc6f8-ddfe-41ec-86ce-2014e3052e4e">

## To replicate 
1. just uncomment the call to `useQuery` on line 33 in `pages/index/tsx` 

[Codesandbox](https://codesandbox.io/p/github/Jezfx/next-algolia/main?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cliiwoki1000b3b6y3gt4s1za%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cliiwoki1000d3b6yz34v1nil%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cliiwoki1000b3b6y3gt4s1za%2522%253A%257B%2522id%2522%253A%2522cliiwoki1000b3b6y3gt4s1za%2522%252C%2522activeTabId%2522%253A%2522cliiwzkst00fz3b6yz84ril5n%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cliiwoki1000a3b6yhuzeq0la%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpages%252Findex.tsx%2522%252C%2522id%2522%253A%2522cliiwzkst00fz3b6yz84ril5n%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cliiwoki1000d3b6yz34v1nil%2522%253A%257B%2522id%2522%253A%2522cliiwoki1000d3b6yz34v1nil%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522cliiwoxe5007s3b6y56h1tx35%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cliiwoz6r00ct3b6ye25rp3iv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cliiwoz6r00ct3b6ye25rp3iv%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

This is a the cloned [next example](https://github.com/algolia/instantsearch/tree/master/examples/react-hooks/next). I've just  added a `QueryClientProvider` in `_app` and also a call to `useQuery` inside of the `Hit` component. 

Please would anyone be able to help, thank you so much!

**Comment by Haroenv** [comment]
For the render done by `getServerState`, you only need to know what InstantSearch widgets are rendered. Any conditionals used by other parts of the apps, hooks, requests etc. that don't change which widgets mount are not needed.

Two workarounds are possible:

1. make a stripped down version of your InstantSearch app, and pass only that to getServerState
2. make the useQuery conditional based on `serverUrl` or another prop that you know is only passed in getServerState

**Reply by Jezfx** [reply]
@Haroenv thank you for your reply! 

From what your saying, I understand that we should ideally only be wrapping Algolia widgets inside of the `InstantSearch` Component. 

What we were hoping to do is wrap the whole `_app` `InstantSearch` and the `InstantSearchSSRProvider` search so that separate components (shop and the sidebar) can both use hooks to pick up things like active refinements (for active filters) and hits for the products etc... 

```
┌───────────────────────────────────────────────────────────────┐
│                         InstantSearch                         │
├───────────────────────────────────────────────────────────────┤
│  ┌──────────────────────────┐ ┌──────────────────────────┐    │
│  │     Filters Sidebar      │ │           Shop           │    │
│  ├──────────────────────────┤ ├──────────────────────────┤    │
│  │                          │ │                          │    │
│  │ useCurrentRefinements()  │ │        useHits()         │    │
│  │                          │ │                          │    │
│  │                          │ │                          │    │
│  └──────────────────────────┘ └──────────────────────────┘    │
└───────────────────────────────────────────────────────────────┘
```

Is that not possible with SSR? 
thanks again! 🙏

**Reply by Haroenv** [reply, marked as answer]
Yes, that is possible, but you'd have to make sure your entire app is renderable when you pass it to getServerState, and thus has no calls to useQuery for example.

What you could do, to still ensure everything working with both the global setup and app to find the server state, is to split up into two different components:

```jsx
// InstantSearch-implementation.tsx

// implement and export all InstantSearch widgets here, so they are colocated
// use this also in your real application

function MockApp() {
  return (
  <InstantSearch>
  {/* render all components using InstantSearch */}
  </InstantSearch>
  )
}

// use <MockApp /> in getServerState
```

**Reply by Jezfx** [reply]
@Haroenv thank you! I understand what you mean. I've updated the Codesandbox with the changes that you suggested and its working now! I'll leave the link below incase it helps anyone else. Many thanks! 

[Codesandbox link](https://codesandbox.io/p/github/Jezfx/next-algolia/main?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cliixq7pl00la3b6yzdyq3aw4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cliiwoki1000d3b6yz34v1nil%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B66.81468324452338%252C33.18531675547662%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cliixq7pl00la3b6yzdyq3aw4%2522%253A%257B%2522id%2522%253A%2522cliixq7pl00la3b6yzdyq3aw4%2522%252C%2522activeTabId%2522%253A%2522cliiwzkst00fz3b6yz84ril5n%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cliiwoki1000a3b6yhuzeq0la%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpages%252Findex.tsx%2522%252C%2522id%2522%253A%2522cliiwzkst00fz3b6yz84ril5n%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpages%252F_app.js%2522%252C%2522id%2522%253A%2522clik2app2007w3b6ym7gjcsck%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cliiwoki1000d3b6yz34v1nil%2522%253A%257B%2522id%2522%253A%2522cliiwoki1000d3b6yz34v1nil%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522cliiwoxe5007s3b6y56h1tx35%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cliiwoz6r00ct3b6ye25rp3iv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cliiwoz6r00ct3b6ye25rp3iv%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
